reference/hoon-expressions/rune/buc.md

```diff
@@ -52,6 +51,84 @@ Irregular: `*` makes `%noun`, `^` makes `%cell`, `?` makes
 
 ## Runes
 
+
+### $| "bucbar"
+
+`[%bsbr p=spec q=hoon]`: structure that satisfies a validator.
+
+##### Expands to
+
+Still need help with this. From hoon.hoon:
+```hoon
+      ::  verify, $|
+      ::
+          {$bsbr *}
+        ^-  hoon
+        ::  push the raw product
+        ::
+        :+  %tsls  relative(mod p.mod)
+        ^-  hoon
+        ::  assert
+        ::
+        :+  %wtbn
+          ::  run the verifier
+          ::
+          [%cnhp [%tsbn $/3 q.mod] $/2]
+        ::  produce verified product
+        ::
+        $/2
+```
```

Comment:
Oof ok. The `$/2` etc expand to `[%$ 2]` etc, and are tree addressing. So, this becomes, using the more-familiar lark syntax:

```hoon
=+  relative(mod p.mod)
?>  (=>(+ q.mod) -)
-
```

In a slightly more readable style, with some shorthand/fake names:

```hoon
=/  =hoon  processed-spec
?>  (=>(+.hoon validator) -.hoon)
-.hoon
```

So, it's evaluating our validator code against the head of the (processed) spec, then calling the resulting function with the tail of the (processed) spec, and asserting that the result is true (as in, good/sane data).

I _think_ the head vs tail of the generated hoon is about getting the bunt vs the type-function itself, but not 100% clear on that. `+relative` just produces a hoon, so idk how this is safe/sane to do.

Once again pinging @joemfb for the details and sanity check. Most of the above I just figured out on the spot. (lmk if you want me to ping someone else going forward)

Reply:
I don't think we want to include the code-generation from $spec to $hoon in these entries, some of it is fairly complicated and subtle. I'm honestly not sure if the "expands to" sections are worth it for molds.

It seems fairly straightforward to say that `$|(a-mold a-validation-gate)` constructs a mold that will apply `a-validation-gate` to the product of `a-mold` when clammed, but less obvious to write out the code that it will use to do so.

That being said, this code is generating a mold-gate from a `$|`. Let's say that the mold is `a` and the validation-gate is `b` (ie, `$|(a b)`. The resulting gate will

- pin the product of clamming with `a` (call this `foo`)
- call `b` (the validation gate) on foo
- assert that the product of `(b foo)` is `&`
- and produce `foo`

So `$|(a b)` is something like

```
|=  x=*
=/  foo  ;;(a x)
?>  (b foo)
b
```